File: cpp/55_链表中环的入口结点.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/
class Solution {
public:
    std::map<ListNode*, int> map;
    ListNode* EntryNodeOfLoop(ListNode* head) {
        while (head != nullptr) {
            if (map.count(head) == 0) {
                map[head] = head->val;
                head = head->next;
            } else {
                return head;
            }
        }
        
        return nullptr;
    }
};

File: python/55_链表中环的入口结点.py
#!/bin/bash/ python3
#-*- coding:utf-8 -*-

# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def EntryNodeOfLoop(self, head):
        # write code here
        tmp = []
        while head != None:
            if tmp.count(head) == 0:
                tmp.append(head)
                head = head.next
            else:
                return head
        
        return None